utils/unminify.js

```diff
@@ -0,0 +1,131 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * @fileoverview
+ * Convert's stacktrace line, column number and file references from minified
+ * to unminified. Caches requests for and source maps once obtained.
+ */
+
+const sourceMap = require('source-map');
+const Request = require('./request');
+const regex = require('./regex');
+const Cache = require('./cache').Cache;
+/** @type {!sourceMap<url, sourceMap>}*/
+const sourceMapConsumerCache = new Cache();
+/** @type {!request<url, Promise>}*/
+const requestCache = new Map();
+
+/**
+ * @param {location} stackLocation
+ * @param {Object} sourceMapConsumer
+ * @return {string} Stack trace line with column, line number and file name
+ * references unminified.
+ */
+function unminifyLine(stackLocation, sourceMapConsumer) {
+  const currentPosition = ':' + stackLocation.lineNumber + ':' +
+      stackLocation.columnNumber;
+  const originalPosition = sourceMapConsumer.originalPositionFor({
+    line: stackLocation.lineNumber,
+    column: stackLocation.columnNumber,
+  });
+  stackLocation.stackTraceLine = stackLocation.stackTraceLine.replace(
+      stackLocation.sourceUrl, originalPosition.source);
+  const originalLocation = ':' + originalPosition.line + ':'
+      + originalPosition.column;
+  return stackLocation.stackTraceLine.replace(currentPosition,
+      originalLocation);
+}
+
+/**
+ * @param {string} url
+ * @return {Promise} Promise that resolves to a source map.
+ */
+function getSourceMapFromNetwork(url) {
+  const reqPromise = new Promise((resolve, reject) => {
+    Request.request(url, function callback(err, _, body) {
+      if (err) {
+        reject(err);
+      } else {
+        try {
+          const sourceMapConsumer = new sourceMap.SourceMapConsumer(
+              JSON.parse(body));
+          requestCache.delete(url);
+          sourceMapConsumerCache.set(url, sourceMapConsumer);
+          resolve(sourceMapConsumer);
+        } catch (e) {
+          reject(e);
+        }
+      }
+    });
+  });
+  requestCache.set(url, reqPromise);
+  return reqPromise;
+}
+
+/**
+ * @param {!Array<!location>} stackLocations
+ * @return {!Array<!Promise>} Array of promises that resolve to source maps
+ */
+function extractSourceMaps(stackLocations) {
+  return stackLocations.map(function(stackLocation) {
+    if (sourceMapConsumerCache.has(stackLocation.sourceMapUrl)) {
+      return Promise.resolve(sourceMapConsumerCache.get(
+          stackLocation.sourceMapUrl));
+    } else if (requestCache.has(stackLocation.sourceMapUrl)) {
+      return requestCache.get(stackLocation.sourceMapUrl);
+    } else {
+     return getSourceMapFromNetwork(stackLocation.sourceMapUrl);
+    }
+  });
+}
+
+/**
+ * @param {string} stackTrace
+ * @return {Promise} Promise that resolves to unminified stack trace.
+ */
+function unminify(stackTrace) {
+  const chromeStackTraceRegex = regex.chromeRegex();
```

Comment:
I tried moving it up and it failed some tests. The fix was to reset last index at the beginning of the function, so I figured why not make it a function level argument?